### PR TITLE
Stabilize prompts [FROM SCRATCH]: fix routing, improve context handling

### PR DESCRIPTION
**Pull request by rafmacalaba**
## Summary

This PR stabilizes the Data360 Chat assistant by fixing a critical bug where **follow-up questions produced empty or invisible responses**. It also improves routing accuracy, conversation context handling, and output formatting.

## Bug Fix: Empty Responses on Follow-up Questions

Follow-up questions (e.g., "Is it important?" after "What is GDP?") caused the entire response to be trapped inside collapsible thinking blocks instead of appearing in the main chat area. Three root causes were identified and fixed:

### 1. Malformed messages sent to routing LLM → `routing.py`

The routing function passed raw conversation history directly to the routing LLM, including multi-part content lists, orphaned `tool_calls` keys, and routing metadata from previous turns. This caused the model to return empty responses, triggering JSON parse errors and incorrectly defaulting to RESEARCH mode.

**Fix:** Added `_simplify_content_for_routing()` to build clean `{role, content}` messages. Filters out tool messages, strips `tool_calls`, converts multi-part content to plain strings, and removes routing metadata.

### 2. `max_tokens` too small for routing model → `routing.py`

`max_tokens=50` was insufficient for the routing model's JSON response format, causing empty output.

**Fix:** Increased to `max_tokens=200`.

### 3. No fallback when `<ANSWER>` token is missing → `stream.py`

In RESEARCH mode, the stream processor expects the LLM to emit an `<ANSWER>` token to switch from thinking to chat mode. When the LLM skipped this token (common on simple follow-ups), all content remained invisible.

**Fix:** Added a fallback at end-of-stream: when the LLM finishes without emitting `<ANSWER>`, the stream processor switches to chat mode and re-emits the buffered content as a regular text part.

### 4. Fixed Router Classification for Visualization Queries
Visualization-related follow-ups (e.g., "Can we create a chart?") were misclassified as DIRECT. Updated routing prompt to route these to RESEARCH.

### 5. Added Conversation Context Handling
Planner was re-fetching data for follow-ups. Added CONVERSATION CONTEXT section to reuse already-retrieved data.

### 6. Improved Source Formatting
Added clearer citation format: `Database — Indicator — methodology note`.

### 7. Restored Number Formatting Examples
Added explicit comma-formatted examples (e.g., `361,751,145,451.597`) to ensure reliable number formatting.

### 8. Strengthened Prompt Instructions
- Reinforced `<ANSWER>` token requirement in RESEARCH prompt
- Added explicit DIRECT mode instructions to never use thinking blocks
- Replaced "Do NOT" with "NEVER" for more emphatic guidance

## Files Changed

| File | Change |
|------|--------|
| `backend/app/ai/routing.py` | Message sanitization, metadata filtering, `max_tokens` bump, better error handling |
| `backend/app/utils/stream.py` | `<ANSWER>` token fallback at end-of-stream |
| `backend/app/ai/prompts.py` | Routing, planner, writer, and direct prompt improvements |
| `backend/app/config.py` | Config updates |

## Testing

- **Manual testing** — verified DIRECT→DIRECT and RESEARCH→DIRECT follow-up sequences render in main chat area

### CODE REVIEW

**Comment by rafmacalaba** [comment]
@avsolatorio  please test locally, it is working smoothly on my end.

**Comment by avsolatorio** [comment]
Hello @rafmacalaba, it's generally stable, but there are some edge cases. 👇🏽

---
1. LLM thinks development is about "software development".

image


---

Inside the thinking stream

2. The response to the second question didn't generate the `<ANSWER>` token. I see there's a special handling in the implementation to copy the content of the thinking part when the LLM doesn't generate the `<ANSWER>` token. But it doesn't take into account the case below.

image

---

3. There's also an instance where it doesn't detect the visualization tool.

image


image

**Comment by rafmacalaba** [comment]
Thank you for testing! @avsolatorio see below responses / fixes

> Hello @rafmacalaba, it's generally stable, but there are some edge cases. 👇🏽
> 
> 1. LLM thinks development is about "software development".
> 
> image
> Inside the thinking stream
> 
> 2. The response to the second question didn't generate the `<ANSWER>` token. I see there's a special handling in the implementation to copy the content of the thinking part when the LLM doesn't generate the `<ANSWER>` token. But it doesn't take into account the case below.
> 
> image
> 3. There's also an instance where it doesn't detect the visualization tool.
> 
> image image

for 1. I modified the prompt to add `-data` in the `development` to avoid confusion, also on the development-related to `development data-related`.

image

2. Not sure how to reproduce, I asked the same thing and it was routed correctly in research mode and it was correctly handoff to the writer (I'll try more examples)

image

3. `There's also an instance where it doesn't detect the visualization tool.`, The chat was routed to direct which has no access to tools that is why, and it is not switching mode.

reproduced and got results (multiple times with "visualize" and "chart" after retrieving data)
image


for #44 
image

**Comment by avsolatorio** [comment]
> 2. Not sure how to reproduce, I asked the same thing and it was routed correctly in research mode and it was correctly handoff to the writer (I'll try more examples)

Hello @rafmacalaba, this seems to happen when the last part of the response is a tool instead of a text response. We can merge this in the meantime and track how frequently this occurs.

**Comment by rafmacalaba** [comment]
> > 2. Not sure how to reproduce, I asked the same thing and it was routed correctly in research mode and it was correctly handoff to the writer (I'll try more examples)
> 
> Hello @rafmacalaba, this seems to happen when the last part of the response is a tool instead of a text response. We can merge this in the meantime and track how frequently this occurs.

This is noted, thanks master! I am thinking on how to evaluate this systematically, any thoughts or suggestions? I saw something like deepeval, I can make a run-up.

**Comment by avsolatorio** [comment]
> > > 2. Not sure how to reproduce, I asked the same thing and it was routed correctly in research mode and it was correctly handoff to the writer (I'll try more examples)
> > 
> > 
> > Hello @rafmacalaba, this seems to happen when the last part of the response is a tool instead of a text response. We can merge this in the meantime and track how frequently this occurs.
> 
> This is noted, thanks master! I am thinking on how to evaluate this systematically, any thoughts or suggestions? I saw something like deepeval, I can make a run-up.

Hello @rafmacalaba , yes! I would tie it back to our MCP evals task.

File: backend/app/utils/stream.py [hunk 1]
```diff
@@ -707,6 +707,44 @@ async def stream_text(
 
             # Handle text end - emit if text was started and stream finished
             if finish_reason == "stop" and text_started and not text_finished:
+                # CRITICAL FALLBACK: If we expected an <ANSWER> token but never
```
Comment (avsolatorio) [inline backend/app/utils/stream.py:710]: We should stress-test this later. The fact that the `<ANSWER>` token is not generated means the system/LLM is failing.

**Review by avsolatorio — APPROVED** [review]
Left one comment, but for later... LGTM! :)
